### PR TITLE
add crash reporting module (WIP)

### PR DESCRIPTION
**Pull request by garbageslam**
update: 
- using "libunwind" seems to generate great stacktraces on linux
- the signal handler registration mechanism does not seem to work consistently. about 50% of the time it is called in case of segfault, and otherwise it seems just to be ignored.
  i have been able to reproduce the ignoring in gdb. i have also been able to observe in gdb that just before the signal handler is ignored, it is also not registered, according to sigaction (using it manually from gdb as described here: 
  http://stackoverflow.com/questions/6892252/how-to-get-process-signal-information-in-gdb
- possibly related to threading? does it need to assign a signal handler for every thread?

### CODE REVIEW

File: src/StackWalker.cpp [hunk 1]
```diff
@@ -0,0 +1,1356 @@
...
+    if (hPsapi == nullptr)
+      return FALSE;
+
+    pEPM = (tEPM) GetProcAddress( hPsapi, "EnumProcessModules" );
+    pGMFNE = (tGMFNE) GetProcAddress( hPsapi, "GetModuleFileNameExA" );
+    pGMBN = (tGMFNE) GetProcAddress( hPsapi, "GetModuleBaseNameA" );
+    pGMI = (tGMI) GetProcAddress( hPsapi, "GetModuleInformation" );
+    if ( (pEPM == nullptr) || (pGMFNE == nullptr) || (pGMBN == nullptr) || (pGMI == nullptr) )
+    {
+      // we couldn�t find all functions
```
Comment (gfgtdf) [inline src/StackWalker.cpp]: invalig utf8 here, GitHub cannot display it.

Comment (cbeck88) [inline src/StackWalker.cpp]: thank you

**Comment by cbeck88** [comment]
Note: I started experimenting with libunwind because, libc only works with ELF format executables and isn't aiming to be generally portable or generally to give good backtraces. Here's an exchange from relevant maintainer:

https://bugzilla.redhat.com/show_bug.cgi?id=169017

> That's a feature.  backtrace_symbols{,_fd} () uses only the dynamic symbol table,
> something that is already loaded and can be looked up easily.
> Hidden symbols aren't in the dynamic symbol table, so they don't show up in
> backtrace_symbols{,_fd} ().
> If that's not enough and either non-allocated symbol table or debugging info
> needs to be analyzed, that's a job of other utilities (like addrline).

> Hello Jakub,
> 
> that's really bad.
> We use the backtrace() function in our application to produce a backtrace in a
> log file, which then later can be used by development to find the real problem.
> It's really unhandy to ask customers to run addrline or such.
>
> ...

> FYI, backtrace () just gives you addresses, you are really talking about
> backtrace_symbols{,_fd} ().  But glibc is certainly not meant to be a kitchen
> sink and even just finding the library, opening it, finding the symbol and string
> tables and looking information there would mean glibc would need to contain
> quite a lot of ELF handling code, probably a copy of libelf.  Plus most of the
> symbol and string tables as well as debug info in our distro are stripped into
> separate files, so it would need to handle that too.  Not to mention
> that using debug info as opposed to just symbol table you can get far better
> info, see e.g. addr2line -i option in rawhide that gives you info about inlined
> functions in the backtrace.
>
> Nothing prevents you from spawning addr2line utility (or eu-addr2line) and
> passing it the addresses from backtrace () and reporting back what it printed.
> And that's the best solution here IMHO.

"Spawning addr2line" seems like a much more complicated solution to me than just using a more powerful lib like libunwind that was apparently meant to generate real backtraces and not just addresses.

**Comment by Pentarctagon** [comment]
Closing as this hasn't been updated in ~5 years.  If anyone decides to come back to this and keep working on it, feel free to re-open.

**Comment by garbageslam** [comment]
yah sorry pentactagon, probly someone should do this some day, but it's not a small amount of work

**Pull request closed without merging**